File: Solution/WebSolution/AssessmentPortal/Microservices/WebAPI/QuestionBankAPI/Controller/QuestionBankController.cs
using System.Diagnostics;
using Microsoft.AspNetCore.Mvc;
using Transflower.TFLAssessment.Entities;
using Transflower.TFLAssessment.Services.Interfaces;


//Controller is now responsible to handle HTTP Requests
namespace Transflower.Assessment.WebAPI.QuestionBankAPI.Controllers;

[ApiController]
[Route("api/questionbank")]
public class QuestionBankController : ControllerBase
{
    private readonly ILogger<QuestionBankController> _logger;
    private readonly IQuestionBankService _svc;

    public QuestionBankController(ILogger<QuestionBankController> logger,IQuestionBankService service)
    {
        _logger = logger;
        _svc=service;
    }

    //http://localhost:5172/api/questionbank/questions
    [HttpGet("questions")]
    public async Task<IActionResult> GetAllQuestions()
    {
        List<QuestionTitle> questions =await _svc.GetAllQuestions();
        _logger.LogInformation("Get all products method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(questions);
    }


    //http://localhost:5172/api/questionbank/questions/2
    [HttpGet("questions/{questionId}")]
    public async Task<IActionResult> GetQuestion(int questionId)
    {
        Question question = await _svc.GetQuestion(questionId);
        return Ok(question);
    }

    [HttpGet("questions/subjects/{subject}/questions/{questionId}")]
    public async Task<IActionResult> GetCriteria(string subject, int questionId)
    {
        string criteria = await _svc.GetCriteria(subject,questionId);
        return Ok(criteria);
    }


    //get questions by subject .
    //http://localhost:5172/api/questionbank/questions/subjects/2
    [HttpGet("questions/subjects/{id}")]
    public async Task<IActionResult> GetQuestionsBySubjects(int id)
    {    
        List<SubjectQuestion> questions =await _svc.GetQuestionsBySubject(id);
        return Ok(questions);
    }


         
    //Get questions by testid .
    [HttpGet("questions/tests/{testId}")]
    public async Task<IActionResult> GetQuestions(int testId)
    {   
        
        List<Question> questions =await _svc.GetQuestions(testId);
        return Ok(questions);
    }

        
        
    //Get questions by subject criteria .
    //http://localhost:5172/api/questionbank/questions/subjects/4/criterias/1
    [HttpGet("questions/subjects/{subjectId}/criterias/{criteriaId}")]
    public async Task<IActionResult> GetQuestionsBySubjectAndCriteria(int subjectId,int criteriaId)
    {   
        List<QuestionDetails> questions =await _svc.GetQuestionsBySubjectAndCriteria(subjectId,criteriaId);
        return Ok(questions);
    }


        
       
    //Update  answer of the question. 
    [HttpPut("question/{id}/updateanswer/{answerKey}")]
    public async Task<IActionResult> UpdateAnswer(int id ,char answerKey)
    {   
        bool status = await _svc.UpdateAnswer(id,answerKey);
        return Ok(status);
    }

        

    //Update question options .
    //http://localhost:5172/api/questionbank/update/options/question/1
    [HttpPut("update/options/question/{id}")]
    public async Task<IActionResult> UpdateQuestionOptions(int id,Question options)
    {

        bool  status = await _svc.UpdateQuestionOptions(id,options);

        return Ok(status);
    }

    //  [HttpPut("update/subjectcriteria/question/{questionId}")]
    //   public IActionResult UpdateSubjectCriteria(int questionId,Question question)
    //   {

    //     bool  status = _svc.UpdateSubjectCriteria(questionId,question);

    //     return Ok(status);
    //   }

    //http://localhost:5172/api/questionbank/question
    [HttpPost("question")]
    public async Task<IActionResult> InsertQuestion(NewQuestion question)
    {
        bool status =await  _svc.InsertQuestion(question);
        return Ok(status);
    }
      
}

File: Solution/WebSolution/Java/AssessmentAPI/src/main/java/com/transflower/tflassessment/controllers/QuestionBankController.java
package com.transflower.tflassessment.controllers;

import java.util.List;
import java.util.concurrent.CompletableFuture;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.transflower.tflassessment.entities.NewQuestion;
import com.transflower.tflassessment.entities.Question;
import com.transflower.tflassessment.entities.QuestionDetails;
import com.transflower.tflassessment.entities.QuestionTitle;
import com.transflower.tflassessment.entities.SubjectQuestion;
import com.transflower.tflassessment.services.QuestionBankService;

@RestController
@RequestMapping("api/questionbank")
public class QuestionBankController {
    
    @Autowired
    private QuestionBankService svc;

    @GetMapping("/questions")
    public CompletableFuture<List<QuestionTitle>> getAllQuestions(){
        return svc.getAllQuestions();
    }

    @GetMapping("questions/{questionId}")
    public CompletableFuture<Question> getQuestion(@PathVariable("questionId") int questionId) {
        return svc.getQuestion(questionId);
    }

    @GetMapping("questions/subjects/{subject}/questions/{questionId}")
    public CompletableFuture<String> getConcept(@PathVariable("subject") String subject,@PathVariable("questionId") int questionId){
        return svc.getConcept(subject, questionId);
    }

    @GetMapping("questions/subjects/{id}")
    public CompletableFuture<List<SubjectQuestion>> getQuestionsBySubject(@PathVariable("id")int id){
        return svc.getQuestionsBySubject(id);
    }

    @GetMapping("questions/tests/{testId}")
    public CompletableFuture<List<Question>> getQuestions(@PathVariable("testId")int testId){
        return svc.getQuestions(testId);
    }

    @GetMapping("questions/subjects/{subjectId}/concepts/{conceptId}")
    public CompletableFuture<List<QuestionDetails>> getQuestionsBySubjectAndConcept(@PathVariable("subjectId") int subjectId,@PathVariable("conceptId")int conceptId){
        return svc.getQuestionsBySubjectAndConcept(subjectId, conceptId);
    }

    @GetMapping("/questions/subjects/concepts")
    public CompletableFuture<List<QuestionDetails>> getQuestionsWithSubjectAndConcept(){
        return svc.getQuestionsWithSubjectAndConcept();
    }

    @PutMapping("/question/{id}/updateanswer/{answerKey}")
    public CompletableFuture<Boolean> updateAnswer(@PathVariable("id")int id,@PathVariable("answerKey") char answerKey){
        return svc.updateAnswer(id, answerKey);
    }

    @PutMapping("/update/options/question/{id}")
    public CompletableFuture<Boolean> updateQuestionOptions(@PathVariable("id")int id,@RequestBody Question option){
        return svc.updateQuestionOptions(id, option);
    }

    @PostMapping("/question")
    public CompletableFuture<Boolean>  insertQuestion(@RequestBody NewQuestion question){
        return svc.insertQuestion(question);
    }    

}